### DISCUSSION TITLE
Fehlende Features?

### DISCUSSION THREAD

**Discussion by Todie99** [FI18-Trainees/vvspy#78] [Ideas]
Derzeit wäre wird die Haltestelle Heerstraße (2593) einseitig nicht angefahren.
Die Fahrten werden (wenn ich nichts übersehen habe) als pünktlich ausgegeben.
In kwargs.get("realtimeStatus", "") wird allerdings (richtigerweise) ausgegeben "DEPARTURE_CANCELLED".

Desweiteren wäre es schön, wenn die Position der Haltestelle im Objekt enthalten wäre. So wäre es möglich an einer Haltestelle nur eine Richtung zu filtern, ohne alle möglichen Ziele angeben zu müssen, die sich im Laufe der Zeit, auch kurzfristig - z.B. aufgrund von Baustellen - ändern können.

**Comment by zaanposni** [comment]
> Derzeit wäre wird die Haltestelle Heerstraße (2593) einseitig nicht angefahren.

Ich kenn mich mit den internen Gegebenheiten der VVS Daten leider auch nicht genau aus.
Was ich rausfinden konnte ist, dass die Verbindungen auf vvs.de ja weiterhin angezegit werden.
![grafik](https://github.com/zaanposni/vvspy/assets/24491035/e3a65327-ce6b-4d0a-9d3a-5f7556dc7fab)
Dementsprechend sind die Daten auch so in Python verfügbar. Ich sehe das erstmal als richtiges Verhalten an.

Worüber man jetzt nachdenken müsste, ist ob es sinnvoll wäre den `realtimeStatus` prominenter zu dokumentieren. Dessen Verwendung muss dann halt von dem jeweiligen Programm (deine Seite) übernommen werden. Ich werd mir das am Wochenende mal genauer anschauen und integrieren.


> Desweiteren wäre es schön, wenn die Position der Haltestelle im Objekt enthalten wäre

Dafür helfen dir vermutlich die "Platform" Properties weiter.
```python
from vvspy import get_departures, get_trips
from vvspy.enums import Station

departures = get_departures(Station.HAUPTBAHNHOF__TIEF, limit=5)

for departure in departures:
    print(f"{departure.serving_line.name} von {departure.platform} ({departure.platform_name})")

trips = get_trips(Station.VAIHINGEN, Station.FEUERBACH, limit=1)

for trip in trips:
    for connection in trip.connections:
        print(f"From {connection.origin.disassembled_name} ({connection.origin.id}) to {connection.destination.disassembled_name} ({connection.destination.id})")
```
Für Trips und Departures sind die unterschiedlich. 
- Du könntest einmal die Strings direkt verwenden. Beispielsweise `platform_name` -> "Gleis 102"
- Oder du nutzt die ID einer Station. Die sind ungefähr so aufgebaut: `de:08111:6118:1:102`. Wobei `6118` für Hauptbahnhof Tief steht und die letzten `102` stehen für Gleis 102. Das müsstest du halt selbst parsen im Gegensatz zu den Strings.

Notiz: Bei `disassembled_name` scheint es einen kleinen Bug zu geben. Das ist moment ein Tuple anstatt nur ein String :/

**Reply by Todie99** [reply]
> > Derzeit wäre wird die Haltestelle Heerstraße (2593) einseitig nicht angefahren.
> 
> Ich kenn mich mit den internen Gegebenheiten der VVS Daten leider auch nicht genau aus. Was ich rausfinden konnte ist, dass die Verbindungen auf vvs.de ja weiterhin angezegit werden.  Dementsprechend sind die Daten auch so in Python verfügbar. Ich sehe das erstmal als richtiges Verhalten an.
> 
> Worüber man jetzt nachdenken müsste, ist ob es sinnvoll wäre den `realtimeStatus` prominenter zu dokumentieren. Dessen Verwendung muss dann halt von dem jeweiligen Programm (deine Seite) übernommen werden. Ich werd mir das am Wochenende mal genauer anschauen und integrieren.

Da bin ich bei Dir. Die Tatsache, daß die Fahrten angezeigt werden empfinde ich auch als richtig. Auch auf der Seite des VVS werden diese im Haltenmonitor angezeigt. Allerdigs jeweils mit dem Hinweis "Ausfall".

https://dfisite.vvs.de/?default_stop=de%3A08111%3A2593&time_offset=3&zoom_factor=1&rows=7&show_messages=1&show_position=0&lpm=400&lang=de&pk_campaign=entwicklertools&pk_kwd=abfahrtstafel

> 
> > Desweiteren wäre es schön, wenn die Position der Haltestelle im Objekt enthalten wäre
> 
> Dafür helfen dir vermutlich die "Platform" Properties weiter.
> 
> ```python
> from vvspy import get_departures, get_trips
> from vvspy.enums import Station
> 
> departures = get_departures(Station.HAUPTBAHNHOF__TIEF, limit=5)
> 
> for departure in departures:
>     print(f"{departure.serving_line.name} von {departure.platform} ({departure.platform_name})")
> 
> trips = get_trips(Station.VAIHINGEN, Station.FEUERBACH, limit=1)
> 
> for trip in trips:
>     for connection in trip.connections:
>         print(f"From {connection.origin.disassembled_name} ({connection.origin.id}) to {connection.destination.disassembled_name} ({connection.destination.id})")
> ```
> 
> Für Trips und Departures sind die unterschiedlich.
> 
>     * Du könntest einmal die Strings direkt verwenden. Beispielsweise `platform_name` -> "Gleis 102"
> 
>     * Oder du nutzt die ID einer Station. Die sind ungefähr so aufgebaut: `de:08111:6118:1:102`. Wobei `6118` für Hauptbahnhof Tief steht und die letzten `102` stehen für Gleis 102. Das müsstest du halt selbst parsen im Gegensatz zu den Strings.
> 
> 
> Notiz: Bei `disassembled_name` scheint es einen kleinen Bug zu geben. Das ist moment ein Tuple anstatt nur ein String :/

Genau, an die Platform-Properties wollte ich ran! Irgendwie hatte ich die bei Dir nicht gefunden und in der Response vom VVS wird das ja für jede einzelne Abfahrt angegeben - ich habe aber von Python auch nicht wirklich Ahnung ... schaue ich mir demnächst nochmal an, vielen Dank :-)
Das Ziel, wo ich hinmöchte: gibt mir alle Fahrten der Haltestelle, die auch tatsächlich stattfinden und zwar nur in eine Richtung.

**Comment by zaanposni** [comment]
@Todie99 

Für Arrivals und Departures gibt es nun (Version `2.2.0`) die Property `cancelled` . Ich vermute, dass sie nicht alle Fälle abfangen wird und manchmal ein false negative ausgibt. Aber schonmal besser als nichts.

Für die `cancelled` Property und dem Thema mit den Platformen habe ich im [readme](https://github.com/zaanposni/vvspy/tree/master?tab=readme-ov-file#examples) Beispiele hinzugefügt. Damit müsstest du dein Projekt auf jeden Fall voranbringen können :)

Falls du noch weitere Themen hast oder du vielleicht schon fertig bist und deine Erfahrungen teilen willst, melde dich gerne hier in den "Discussions". Davon profitieren alle :)

**Comment by Todie99** [comment]
Wow, seeeeehr cool!
Vielen Dank. Eigentlich wollte ich mich da am WE mal dran machen ... wird wohl eher das Nächste ... :-)
